gstd/src/common/handlers.rs

```diff
@@ -25,50 +25,209 @@
+            /// Max amount of bytes allowed to be thrown as string explanation
+            /// of the error.
+            pub const TRIMMED_MAX_LEN: usize = 1024; //TODO: do not duplicate `gear_core::str::TRIMMED_MAX_LEN`
+            /// This prefix is used to print debug message:
+            /// `debug!("panic occurred: {msg}")`.
+            pub const PANIC_OCCURRED: &str = "panic occurred: ";
+            /// This prefix is used by `impl Display for PanicInfo<'_>`.
+            #[cfg(not(feature = "panic-messages"))]
+            pub const PANICKED_AT: &str = "panicked at ";
+        }
+
+        #[cfg(feature = "debug")]
+        use constants::*;
+
+        /// Panic handler for nightly Rust.
+        #[cfg(feature = "debug")]
         #[cfg(feature = "panic-messages")]
-        let message = panic_info.message();
-
-        let msg = match (message, panic_info.location()) {
-            (Some(msg), Some(loc)) => format!(
-                "'{:?}', {}:{}:{}",
-                msg,
-                loc.file(),
-                loc.line(),
-                loc.column()
-            ),
-            (Some(msg), None) => format!("'{msg:?}'"),
-            (None, Some(loc)) => {
-                format!("{}:{}:{}", loc.file(), loc.line(), loc.column())
+        #[panic_handler]
+        pub fn panic(panic_info: &PanicInfo) -> ! {
+            use crate::prelude::{fmt::Write, mem::MaybeUninit, str};
+            use arrayvec::ArrayString;
+
+            let option = panic_info.message().zip(panic_info.location());
+            let (message, location) = unsafe { option.unwrap_unchecked() };
+
+            fn itoa_u32(buffer: &mut [MaybeUninit<u8>; 10], mut n: u32) -> &str {
+                let mut idx = buffer.len();
+                loop {
+                    idx -= 1;
+                    unsafe { buffer.get_unchecked_mut(idx) }.write((n % 10) as u8 + b'0');
+                    n /= 10;
+                    if n == 0 {
+                        break;
+                    }
+                }
+                unsafe {
+                    str::from_utf8_unchecked(
+                        &*(buffer.get_unchecked(idx..) as *const [_] as *const _),
+                    )
+                }
+            }
+
+            let mut debug_msg = ArrayString::<{ PANIC_OCCURRED.len() + TRIMMED_MAX_LEN }>::new();
+            let mut buffer = unsafe { MaybeUninit::uninit().assume_init() };
+
+            let _ = debug_msg.try_push_str(PANIC_OCCURRED);
+            let _ = debug_msg.try_push_str("'");
+            let _ = write!(&mut debug_msg, "{message}");
+            let _ = debug_msg.try_push_str("', ");
+            let _ = debug_msg.try_push_str(location.file());
+            let _ = debug_msg.try_push_str(":");
+            let _ = debug_msg.try_push_str(itoa_u32(&mut buffer, location.line()));
+            let _ = debug_msg.try_push_str(":");
+            let _ = debug_msg.try_push_str(itoa_u32(&mut buffer, location.column()));
```

Comment:
Can be replaced with `write!()`

Reply:
`impl Display for u32` costs 2,5 KiB in binary VS `itoa_u32` 82 bytes

Reply:
So we can leave `itoa_u32` but use still use macro

Reply:
In this case, the size of the binary file will become larger (`2_273 -> `3_824`). Also you can't reuse `buffer` due to taking the mutable reference twice.
```rust
let mut buffer = [MaybeUninit::uninit(); 10];

let _ = write!(
    &mut debug_msg,
    "{PANIC_OCCURRED}'{message}', {}:{}:{}",
    location.file(),
    itoa_u32(&mut buffer, location.line()), //&mut #1
    itoa_u32(&mut buffer, location.column()) //&mut #2
);
```

Reply:
66be25db10710945d5eb639e0b7a9089a8ce8647